### ISSUE TITLE
A question regarding discriminator

### ISSUE THREAD

**Issue by iLampard** [da03/Residual-EBM#1]
Hi,  thanks for sharing the code, it is very nicely organized and easy to understand.

I just have two trivial questions
1) in training discriminators, the labels for real logits is 0 while those for the fake is set to 1. is there any special reason for doing this? 

`` 

      real_logits = self.forward(real_text)
      loss = self.loss_function(real_logits, 
                                torch.LongTensor(real_text.size(1)).fill_(0).to(device))
      total_loss += loss.item()
      # Compute cross entropy loss
      num_correct += real_logits[:,0].ge(real_logits[:,1]).float().sum().item()

      fake_logits = self.forward(fake_text)
      loss = self.loss_function(fake_logits, 
                                torch.LongTensor(fake_text.size(1)).fill_(1).to(device))

      

2) when computing the scores, you use the difference of last dim of logits. I guess this corresponds to the way you compute entropy loss above, where logits[:, 0] refers to the real logits and logits[:, 1] refers to the fake logits, right?  

another questions is why not just use logits[:, 0] as the scores? 

`

     logits = discriminator.forward(text)

      l = logits[:, 0] - logits[:, 1]

      l = l.item()

      scores.append(l)
`

Thanks a lot in advance

**Comment by da03** [comment]
Sorry for the delay... The short answer is there are multiple ways of parameterizing. 

The long answer:
The ids of labels are set arbitrarily: I used 0 for real labels, and 1 for fake labels, so the scores (negative energy) are logits[:,0]-logits[:,1], but it'd be the same if we flip the labels and then use logits[:, 1]-logits[:,0] as the scores.
Another equivalent way is to use a sigmoid loss, and then we no longer need to use two numbers one for label 0 and another for label 1, instead one number is enough, and we just need to take sigmoid to convert the logit to the probability of label 1 (or 0).
